File: desktop/src/main/java/com/securetransfer/desktop/network/FileTransferClient.java
package com.securetransfer.desktop.network;

import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;

public class FileTransferClient {
    private static final Logger logger = LoggerFactory.getLogger(FileTransferClient.class);
    
    private final OkHttpClient client;
    
    public FileTransferClient() {
        this.client = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(0, TimeUnit.SECONDS) // No timeout for file uploads
                .readTimeout(0, TimeUnit.SECONDS)  // No timeout for file downloads
                .build();
    }
    
    public CompletableFuture<String> sendFile(String targetHost, int targetPort, File file, 
                                            BiConsumer<Long, Long> progressCallback) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                logger.info("Sending file {} to {}:{}", file.getName(), targetHost, targetPort);
                
                RequestBody requestBody = RequestBody.create(file, MediaType.get("application/octet-stream"));
                ProgressRequestBody progressRequestBody = new ProgressRequestBody(requestBody, progressCallback);
                
                Request request = new Request.Builder()
                        .url("http://" + targetHost + ":" + targetPort + "/upload")
                        .post(progressRequestBody)
                        .addHeader("Content-Disposition", "attachment; filename=\"" + file.getName() + "\"")
                        .build();
                
                try (Response response = client.newCall(request).execute()) {
                    if (response.isSuccessful()) {
                        logger.info("File sent successfully: {}", file.getName());
                        return "File sent successfully";
                    } else {
                        String error = "Transfer failed: " + response.code() + " " + response.message();
                        logger.error(error);
                        throw new RuntimeException(error);
                    }
                }
            } catch (Exception e) {
                logger.error("Error sending file", e);
                throw new RuntimeException(e);
            }
        });
    }
    
    public CompletableFuture<String> sendText(String targetHost, int targetPort, String text) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                logger.info("Sending text to {}:{}", targetHost, targetPort);
                
                RequestBody requestBody = RequestBody.create(text, MediaType.get("text/plain"));
                
                Request request = new Request.Builder()
                        .url("http://" + targetHost + ":" + targetPort + "/text")
                        .post(requestBody)
                        .build();
                
                try (Response response = client.newCall(request).execute()) {
                    if (response.isSuccessful()) {
                        logger.info("Text sent successfully");
                        return "Text sent successfully";
                    } else {
                        String error = "Transfer failed: " + response.code() + " " + response.message();
                        logger.error(error);
                        throw new RuntimeException(error);
                    }
                }
            } catch (Exception e) {
                logger.error("Error sending text", e);
                throw new RuntimeException(e);
            }
        });
    }
    
    public CompletableFuture<Boolean> pingDevice(String host, int port) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                Request request = new Request.Builder()
                        .url("http://" + host + ":" + port + "/ping")
                        .get()
                        .build();
                
                try (Response response = client.newCall(request).execute()) {
                    return response.isSuccessful() && "pong".equals(response.body().string());
                }
            } catch (Exception e) {
                logger.debug("Ping failed for {}:{}", host, port, e);
                return false;
            }
        });
    }
    
    private static class ProgressRequestBody extends RequestBody {
        private final RequestBody delegate;
        private final BiConsumer<Long, Long> progressListener;
        
        public ProgressRequestBody(RequestBody delegate, BiConsumer<Long, Long> progressListener) {
            this.delegate = delegate;
            this.progressListener = progressListener;
        }
        
        @Override
        public MediaType contentType() {
            return delegate.contentType();
        }
        
        @Override
        public long contentLength() throws IOException {
            return delegate.contentLength();
        }
        
        @Override
        public void writeTo(okio.BufferedSink sink) throws IOException {
            ProgressSink progressSink = new ProgressSink(sink, contentLength(), progressListener);
            okio.BufferedSink bufferedSink = okio.Okio.buffer(progressSink);
            delegate.writeTo(bufferedSink);
            bufferedSink.flush();
        }
    }
    
    private static class ProgressSink extends okio.ForwardingSink {
        private final long contentLength;
        private final BiConsumer<Long, Long> progressListener;
        private long bytesWritten = 0L;
        
        public ProgressSink(okio.Sink delegate, long contentLength, BiConsumer<Long, Long> progressListener) {
            super(delegate);
            this.contentLength = contentLength;
            this.progressListener = progressListener;
        }
        
        @Override
        public void write(okio.Buffer source, long byteCount) throws IOException {
            super.write(source, byteCount);
            bytesWritten += byteCount;
            progressListener.accept(bytesWritten, contentLength);
        }
    }
}

File: android/src/main/java/com/securetransfer/android/network/FileTransferClient.kt
package com.securetransfer.android.network

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

class FileTransferClient(private val context: Context) {
    private val logger = LoggerFactory.getLogger(FileTransferClient::class.java)
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(0, TimeUnit.SECONDS) // No timeout for file uploads
        .readTimeout(0, TimeUnit.SECONDS)  // No timeout for file downloads
        .build()
    
    private val _transferProgress = MutableStateFlow<TransferProgress?>(null)
    val transferProgress: StateFlow<TransferProgress?> = _transferProgress
    
    suspend fun sendFile(
        targetHost: String,
        targetPort: Int,
        fileUri: Uri,
        fileName: String
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            logger.info("Sending file $fileName to $targetHost:$targetPort")
            
            val inputStream = context.contentResolver.openInputStream(fileUri)
                ?: return@withContext Result.failure(Exception("Cannot open file"))
            
            val tempFile = File(context.cacheDir, fileName)
            tempFile.outputStream().use { output ->
                inputStream.use { input ->
                    input.copyTo(output)
                }
            }
            
            val requestBody = tempFile.asRequestBody("application/octet-stream".toMediaType())
            val progressRequestBody = ProgressRequestBody(requestBody) { bytesWritten, contentLength ->
                _transferProgress.value = TransferProgress(
                    fileName = fileName,
                    bytesTransferred = bytesWritten,
                    totalBytes = contentLength,
                    isComplete = bytesWritten >= contentLength
                )
            }
            
            val request = Request.Builder()
                .url("http://$targetHost:$targetPort/upload")
                .post(progressRequestBody)
                .addHeader("Content-Disposition", "attachment; filename=\"$fileName\"")
                .build()
            
            val response = client.newCall(request).execute()
            
            tempFile.delete() // Clean up temp file
            
            if (response.isSuccessful) {
                logger.info("File sent successfully: $fileName")
                Result.success("File sent successfully")
            } else {
                logger.error("File send failed: ${response.code} ${response.message}")
                Result.failure(Exception("Transfer failed: ${response.code} ${response.message}"))
            }
        } catch (e: Exception) {
            logger.error("Error sending file", e)
            Result.failure(e)
        }
    }
    
    suspend fun sendText(
        targetHost: String,
        targetPort: Int,
        text: String
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            logger.info("Sending text to $targetHost:$targetPort")
            
            val requestBody = RequestBody.create("text/plain".toMediaType(), text)
            
            val request = Request.Builder()
                .url("http://$targetHost:$targetPort/text")
                .post(requestBody)
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                logger.info("Text sent successfully")
                Result.success("Text sent successfully")
            } else {
                logger.error("Text send failed: ${response.code} ${response.message}")
                Result.failure(Exception("Transfer failed: ${response.code} ${response.message}"))
            }
        } catch (e: Exception) {
            logger.error("Error sending text", e)
            Result.failure(e)
        }
    }
    
    suspend fun pingDevice(host: String, port: Int): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("http://$host:$port/ping")
                .get()
                .build()
            
            val response = client.newCall(request).execute()
            response.isSuccessful && response.body?.string() == "pong"
        } catch (e: Exception) {
            logger.debug("Ping failed for $host:$port", e)
            false
        }
    }
    
    data class TransferProgress(
        val fileName: String,
        val bytesTransferred: Long,
        val totalBytes: Long,
        val isComplete: Boolean
    ) {
        val progressPercentage: Float
            get() = if (totalBytes > 0) (bytesTransferred.toFloat() / totalBytes.toFloat()) * 100f else 0f
    }
    
    private class ProgressRequestBody(
        private val delegate: RequestBody,
        private val progressListener: (bytesWritten: Long, contentLength: Long) -> Unit
    ) : RequestBody() {
        
        override fun contentType(): MediaType? = delegate.contentType()
        
        override fun contentLength(): Long = delegate.contentLength()
        
        override fun writeTo(sink: okio.BufferedSink) {
            val progressSink = ProgressSink(sink, contentLength(), progressListener)
            val bufferedSink = okio.buffer(progressSink)
            delegate.writeTo(bufferedSink)
            bufferedSink.flush()
        }
    }
    
    private class ProgressSink(
        private val delegate: okio.Sink,
        private val contentLength: Long,
        private val progressListener: (bytesWritten: Long, contentLength: Long) -> Unit
    ) : okio.ForwardingSink(delegate) {
        
        private var bytesWritten = 0L
        
        override fun write(source: okio.Buffer, byteCount: Long) {
            super.write(source, byteCount)
            bytesWritten += byteCount
            progressListener(bytesWritten, contentLength)
        }
    }
}
